### PR TITLE
Q001, Q002, Q003, and Q004

### LINKED ISSUE #184: Validation (2017): Q001 (Loan Amount)
**Edit #:** Q001
**Type:** Quality
**Category:** Loan Application Register (only)
**Column:** Loan Amount

If loan amount and income are numeric and > 0 and the loan amount is ≥ $1,000 ($1 million), then loan amount should be < 5 times the income.

Additional resources:
* [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
* [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
* [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### LINKED ISSUE #185: Validation (2017): Q002 (Loan Amount)
**Edit #:** Q002
**Type:** Quality
**Category:** Loan Application Register (only)
**Column:** Loan Amount

If property type =1 and income is numeric and ≤ $200 thousand, then loan amount should be < $2 million.

Additional resources:
* [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
* [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
* [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### LINKED ISSUE #186: Validation (2017): Q003 (Loan Amount)
**Edit #:** Q003
**Type:** Quality
**Category:** Loan Application Register (only)
**Column:** Loan Amount

If loan type = 2 and property type = 1 or 2, then loan amount should be ≤ $626 ($626 thousand).

Additional resources:
* [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
* [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
* [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### LINKED ISSUE #187: Validation (2017): Q004 (Loan Amount)
**Edit #:** Q004
**Type:** Quality
**Category:** Loan Application Register (only)
**Column:** Loan Amount

If loan type = 3 and property type = 1 or 2, then loan amount should be ≤ $1,050 ($1,050 thousands).

Additional resources:
* [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
* [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
* [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### PR DESCRIPTION
**Pull request by kgudel**
Closes #184 
Closes #185 
Closes #186 
Closes #187

### CODE REVIEW

File: validation/src/main/resources/application.conf [hunk 1]
```diff
@@ -1,4 +1,24 @@
 hmda.validation.quality {
+  Q001 {
+    loan.amount=1000
+    loan.amount=${?Q001_LOAN_AMOUNT}
+    incomeMultiplier=5
+    incomeMultiplier=${?Q001_incomeMultiplier}
```
Comment (schbetsy) [inline validation/src/main/resources/application.conf]: Naming convention for environment variables would be `Q001_INCOME_MULTIPLIER`

... though I'm curious whether the multiplier is likely to change in the next few years. @lesgou or @debseidner , when edits like [Q001](https://github.com/cfpb/hmda-platform/issues/184) say "loan amount should be < 5 times the income", is the 5 likely to change in future years?

Comment (debseidner) [inline validation/src/main/resources/application.conf]: @schbetsy Short answer, it could (i.e. yes) change. Some of these limits are tied to loan rules and others to cut-offs that seem like they would be quality issues so we could change them in the future.

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q001Spec.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      newLar.mustPass
+    }
+  }
+
+  val relevantIncome: Gen[Int] = Gen.choose(1000, 100000)
+  val validMultiplier: Gen[Int] = Gen.choose(1, multiplier - 1)
+
+  property(s"Valid when loan less than $multiplier times income") {
+    forAll(larGen, relevantIncome, validMultiplier) { (lar, x, m) =>
+      val newLoan = lar.loan.copy(amount = x * m)
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q001Spec.scala]: I'd like to see tests for the edge cases--the ones that make this edit barely pass or barely fail.

e.g. `loan = income * 4.99` (or `loan = (income * 5) - 1`)should pass, but `loan = income * 5` should fail

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q001Spec.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  property("Valid whenever Income equal or less than 0") {
+    forAll(larGen, irrelevantIncome) { (lar, x) =>
+      val newApplicant = lar.applicant.copy(income = x.toString)
+      val newLar = lar.copy(applicant = newApplicant)
+      newLar.mustPass
+    }
+  }
+
+  val relevantIncome: Gen[Int] = Gen.choose(1000, 100000)
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q001Spec.scala]: More realistic incomes would be 1-1000 (since it's represented in thousands)

**Comment by schbetsy** [comment]
Q001 wants one more spec, then LGTM.

File: validation/src/main/scala/hmda/validation/rules/lar/quality/Q004.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import hmda.validation.dsl.PredicateSyntax._
+
+object Q004 extends EditCheck[LoanApplicationRegister] {
+  override def apply(lar: LoanApplicationRegister): Result = {
+
+    val config = ConfigFactory.load()
+    val loanAmount = config.getInt("hmda.validation.quality.Q004.loan.amount")
+
+    when((lar.loan.loanType is equalTo(3) and
+      (lar.loan.propertyType is containedIn(List(1, 2))))) {
```
Comment (lgdean) [inline validation/src/main/scala/hmda/validation/rules/lar/quality/Q004.scala]: My IDE thinks there are unnecessary parens here, and I tend to agree with it.
Also, same comment about a weird git history as elsewhere.  (`git log --follow` to see what I mean.)

**Comment by lgdean** [comment]
OK, cool with parens now. (I'd also have been cool with moving one of the closing parens from line 17 up to line 16.)
I'll look into the git history thing some more, because it is the sort of thing that will make maintenance more confusing, but if I haven't said more by tomorrow, don't let it block merging this PR.

File: validation/src/main/scala/hmda/validation/rules/lar/quality/Q003.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+object Q003 extends EditCheck[LoanApplicationRegister] {
+  override def apply(lar: LoanApplicationRegister): Result = {
+
+    val config = ConfigFactory.load()
+    val loanAmount = config.getInt("hmda.validation.quality.Q003.loan.amount")
+
+    when(
+      (lar.loan.loanType is equalTo(2)) and
+        (lar.loan.propertyType is containedIn(List(1, 2)))
```
Comment (nickgrippin) [inline validation/src/main/scala/hmda/validation/rules/lar/quality/Q003.scala]: Can you switch this over to `oneOf`?

File: validation/src/main/scala/hmda/validation/rules/lar/quality/Q004.scala [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import hmda.validation.dsl.PredicateSyntax._
+
+object Q004 extends EditCheck[LoanApplicationRegister] {
+  override def apply(lar: LoanApplicationRegister): Result = {
+
+    val config = ConfigFactory.load()
+    val loanAmount = config.getInt("hmda.validation.quality.Q004.loan.amount")
+
+    when(lar.loan.loanType is equalTo(3) and
+      (lar.loan.propertyType is containedIn(List(1, 2)))) {
```
Comment (nickgrippin) [inline validation/src/main/scala/hmda/validation/rules/lar/quality/Q004.scala]: Same `oneOf` comment here

File: validation/src/main/scala/hmda/validation/rules/lar/quality/Q003.scala (hunk 5 above)
Comment (kgudel) [inline validation/src/main/scala/hmda/validation/rules/lar/quality/Q003.scala]: Oh sure, will do!

**Comment by nickgrippin** [comment]
LGTM!
